dandi/core/digests/dandietag.py

```diff
@@ -0,0 +1,169 @@
+# Derived from <https://github.com/girder/django-s3-file-field/blob/master/
+# s3_file_field/_multipart.py>, copyright Kitware, Inc. <kitware@kitware.com>
+# under the Apache 2.0 license
+
+from dataclasses import dataclass
+from hashlib import md5
+import math
+import os
+from typing import Iterator, List, NamedTuple, Optional, Union
+
+
+def mb(bytes_size: int) -> int:
+    return bytes_size * 2 ** 20
+
+
+def gb(bytes_size: int) -> int:
+    return bytes_size * 2 ** 30
+
+
+def tb(bytes_size: int) -> int:
+    return bytes_size * 2 ** 40
+
+
+class Part(NamedTuple):
+    number: int
+    offset: int
+    size: int
+
+
+@dataclass
+class PartGenerator:
+    part_qty: int
+    initial_part_size: int
+    final_part_size: int
+
+    # S3 multipart limits: https://docs.aws.amazon.com/AmazonS3/latest/dev/qfacts.html
+    # 10k is the maximum number of allowed parts allowed by S3
+    MAX_PARTS = 10_000
+    # 5MB is the minimum part size allowed by S3
+    MIN_PART_SIZE = mb(5)
+    # 5GB is the maximum part size allowed by S3
+    MAX_PART_SIZE = gb(5)
+
+    @classmethod
+    def for_file_size(cls, file_size: int) -> "PartGenerator":
+        """Method to calculate sequential part sizes given a file size"""
+        part_size = mb(64)
+
+        if file_size > tb(5):
+            raise ValueError("File is larger than the S3 maximum object size.")
+
+        if math.ceil(file_size / part_size) >= cls.MAX_PARTS:
+            part_size = math.ceil(file_size / cls.MAX_PARTS)
+
+        if part_size < cls.MIN_PART_SIZE:
+            part_size = cls.MIN_PART_SIZE
+
+        if part_size > cls.MAX_PART_SIZE:
+            part_size = cls.MAX_PART_SIZE
```

Comment:
The above two conditions are actually unreachable.  `part_size` will always be at least 64 MiB, and the only way for `MAX_PART_SIZE` to be exceeded is for the file size to be greater than `MAX_PART_SIZE * MAX_PARTS`, which is greater than 5 TiB, the maximum file size.  I believe these lines are only covered in the original code because the tests monkey-patch the starting value of `part_size`.

Reply:
How do you recommend I do a sample file upload, seeing as the CLI hasn't been updated for the new upload API yet?

Reply:
indeed gets tricky/chicken-egg.  Could postpone and get those while RFing `upload` or just initiating upload manually via API directly

Reply:
I looked at the ETags for some zero-byte entries in s3://dandiarchive.  The only such entries were directories, but they both had ETags of `d41d8cd98f00b204e9800998ecf8427e`, which is the expected hash but without the part quantity (which I think S3 leaves off if there's only one part).

Reply:
re 0 or one part uploads: I would not rely on what you would see in `dandiarchive` bucket since it was uploaded possibly using a different strategy (thus might not even be "always in parts").  A fresh upload should be done to figure it out "for sure".

re unreachable code/conditions -- I would then either remove them or just convert into asserts (to make sure we do not somehow beef up part_size and then fail to acknowlede that down in the code or smth like that)